contracts/StructHash.cairo

```diff
@@ -0,0 +1,114 @@
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.cairo.common.hash_state import (
+    HashState, hash_finalize, hash_init, hash_update, hash_update_single)
+from starkware.cairo.common.registers import get_fp_and_pc
+
+struct StarkNet_Domain:
+    member name : felt
+    member version : felt
+    member chain_id : felt
+end
+
+struct Person:
+    member name : felt
+    member wallet : felt
+end
+
+struct Mail:
+    member from2 : Person
+    member to : Person
+    member contents : felt
+end
+
+# H('StarkNetDomain(felt name,felt version,felt chainId)')
+const STARKNET_DOMAIN_TYPE_HASH = 0x1ac84eab85ae7c91085e6143344d685aba3353d2029a7b1eccf46794249a7
+
+# H('Mail(Person from,Person to,felt contents)Person(felt name,felt wallet)')
+const MAIL_TYPE_HASH = 0x3e42cd1b168ba9dde92479b909ce5133824c58a6b2030aa9fbb698095831ed0
+
+# H('Person(felt name,felt wallet)')
+const PERSON_TYPE_HASH = 0x205728d5ea4ba1cf7fefb11ad4f0bd1eb6f498df394993afb37a89b354dd476
+
+# takes StarkNetDomain and returns its struct_hash
+func hashDomain{hash_ptr : HashBuiltin*}(domain : StarkNet_Domain*) -> (hash : felt):
+    let (hash_state : HashState*) = hash_init()
+    let (hash_state) = hash_update_single(hash_state_ptr=hash_state, item=STARKNET_DOMAIN_TYPE_HASH)
+    let (hash_state) = hash_update_single(hash_state_ptr=hash_state, item=domain.name)
+    let (hash_state) = hash_update_single(hash_state_ptr=hash_state, item=domain.version)
+    let (hash_state) = hash_update_single(hash_state_ptr=hash_state, item=domain.chain_id)
+    let (hash : felt) = hash_finalize(hash_state_ptr=hash_state)
+    return (hash=hash)
+end
+
+# takes Person and returns its struct_hash
+func hashPerson{hash_ptr : HashBuiltin*}(person : Person*) -> (hash : felt):
+    let (hash_state : HashState*) = hash_init()
+    let (hash_state) = hash_update_single(hash_state_ptr=hash_state, item=PERSON_TYPE_HASH)
+    let (hash_state) = hash_update_single(hash_state_ptr=hash_state, item=person.name)
+    let (hash_state) = hash_update_single(hash_state_ptr=hash_state, item=person.wallet)
+    let (hash : felt) = hash_finalize(hash_state_ptr=hash_state)
+    return (hash=hash)
+end
+
+# takes Mail and returns its struct_hash
+func hashMail{hash_ptr : HashBuiltin*}(mail : Mail*) -> (hash : felt):
+    let (hash_state : HashState*) = hash_init()
+    let (hash_state) = hash_update_single(hash_state_ptr=hash_state, item=MAIL_TYPE_HASH)
+    let (personFromHash : felt) = hashPerson(person=&mail.from2)
+    let (hash_state) = hash_update_single(hash_state_ptr=hash_state, item=personFromHash)
+    let (personToHash : felt) = hashPerson(person=&mail.to)
+    let (hash_state) = hash_update_single(hash_state_ptr=hash_state, item=personToHash)
+    let (hash_state) = hash_update_single(hash_state_ptr=hash_state, item=mail.contents)
+    let (hash : felt) = hash_finalize(hash_state_ptr=hash_state)
+    return (hash=hash)
+end
+
+func hashMessage{hash_ptr : HashBuiltin*}(
```

Comment:
The other contracts of the repo use a different function declaration convention which I would suggest to follow for consistency (and readability).

```
func hashMessage{
        hash_ptr : HashBuiltin*
    } (
        domain : StarkNet_Domain,
        mail : Mail,
        account : felt
    ) -> (hash: felt):
```

Reply:
I'm using the auto formatter used by vscode-cairo plugin. Seems like the method definition is too short to go multiline